src/NumericalAlgorithms/Interpolation/InterpolationTargetDetail.hpp

```diff
@@ -0,0 +1,327 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <memory>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/VariablesTag.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+/// \cond
+// IWYU pragma: no_forward_declare db::DataBox
+namespace intrp {
+template <class Metavariables>
+struct Interpolator;
+template <typename Metavariables, typename InterpolationTargetTag>
+class InterpolationTarget;
+namespace Actions {
+template <typename InterpolationTargetTag>
+struct CleanUpInterpolator;
+}  // namespace Actions
+namespace Tags {
+template <typename TemporalId>
+struct IndicesOfFilledInterpPoints;
+template <typename TemporalId>
+struct CompletedTemporalIds;
+template <typename TemporalId>
+struct TemporalIds;
+}  // namespace Tags
+}  // namespace intrp
+template <typename TagsList>
+struct Variables;
+/// \endcond
+
+namespace intrp {
+
+namespace InterpolationTarget_detail {
+
+// apply_callback accomplishes the overload for the
+// two signatures of callback functions.
+// Uses SFINAE on return type.
+template <typename T, typename DbTags, typename Metavariables,
+          typename TemporalId>
+auto apply_callback(
+    const gsl::not_null<db::DataBox<DbTags>*> box,
+    const gsl::not_null<Parallel::ConstGlobalCache<Metavariables>*> cache,
+    const TemporalId& temporal_id) noexcept
+    -> decltype(T::post_interpolation_callback::apply(box, cache, temporal_id),
+                bool()) {
+  return T::post_interpolation_callback::apply(box, cache, temporal_id);
+}
+
+template <typename T, typename DbTags, typename Metavariables,
+          typename TemporalId>
+auto apply_callback(
+    const gsl::not_null<db::DataBox<DbTags>*> box,
+    const gsl::not_null<Parallel::ConstGlobalCache<Metavariables>*> cache,
+    const TemporalId& temporal_id) noexcept
+    -> decltype(T::post_interpolation_callback::apply(*box, *cache,
+                                                      temporal_id),
+                bool()) {
+  T::post_interpolation_callback::apply(*box, *cache, temporal_id);
+  // For the simpler callback function, we will always clean up volume data, so
+  // we return true here.
+  return true;
+}
+
+// type trait testing if the class has a fill_invalid_points_with member.
```

Comment:
can you use one of the macros e.g. `Utilities/TypeTraits/CreateHasStaticMemberVariable.hpp`